custom/icds_reports/models/manager.py

```diff
@@ -0,0 +1,20 @@
+from __future__ import absolute_import, unicode_literals
+
+import uuid
+
+from django.db.models.manager import BaseManager
+from django.db.models.query import QuerySet
+
+from corehq.toggles import ICDS_COMPARE_QUERIES_AGAINST_CITUS, NAMESPACE_OTHER
+
+
+class CitusComparisonQuerySet(QuerySet):
+    def _fetch_all(self):
+        from custom.icds_reports.tasks import run_citus_experiment_raw_sql
+        if ICDS_COMPARE_QUERIES_AGAINST_CITUS.enabled(uuid.uuid4().hex, NAMESPACE_OTHER):
+            run_citus_experiment_raw_sql.delay(str(self.query), data_source=self.model.__name__)
+        super(CitusComparisonQuerySet, self)._fetch_all()
```

Comment:
Do you also need to override `count()`? It executes a `SELECT COUNT(*)` query if the rows for the queryset are not yet loaded. I assume you will not need to override `update`, `_update` and `_insert`, but they also execute queries.

Reply:
Looks like this executes the query immediately here, and then also again on the same database in the task as part of the experiment.

https://github.com/dimagi/commcare-hq/blob/7b91e11db1f71cd0ccf2ed5f783a9c46c65356d0/custom/icds_reports/tasks.py#L1192

I think this may dramatically affect the apparent performance of the "control" query in the experiment since PostgreSQL will have loaded indexes and other caches into RAM, which could make the second run of the same query much faster than the first run. Not sure how hard it would be, but a better approach may be to time this query and send that to the task for comparison in the experiment rather than the time of the control query.

Reply:
This is similar to what @sravfeyn mentioned (though one focused on timing vs results), so copying his comment here and responding in one place.

> Doesn't this already get executed in SqlData itself, why not capture results from that instead of doing a repeat calculation?

The timing issue and performance issues are known as this has also been observed in our UCRLaboratorExperiement (where this code was inspired from). The timing (at least from my perspective) should only be used as a guide, not as final word. We want to check correct-ness first and then tune. Also if we run into any load issues, we want to be able to de-prioritize these so we could turn off hte queue and have them run at a later time.

Because we are delaying the task, we don't know at what time it will run and if the underlying tables will have changed in the meantime, so the task will need to re-run the query to get most up to date data. Also because we have data that's retrieved from sqlalchemy (SqlData) and django, but running the executes the same way (via raw sql), sending the timing from the original run would include slightly different python code in its timing

Reply:
> Do you also need to override count()? It executes a SELECT COUNT(*) query if the rows for the queryset are not yet loaded. I assume you will not need to override update, _update and _insert, but they also execute queries.

intentionally didn't override update/insert. I also did not override count because it looks like all instances where we use `count`, we also loop through all the data, so they would be redudant. for example: https://github.com/dimagi/commcare-hq/blob/0fc7a7962579cef554b30028e56923e2d05903c1/custom/icds_reports/reports/awc_reports.py#L1110 but in taht same function https://github.com/dimagi/commcare-hq/blob/0fc7a7962579cef554b30028e56923e2d05903c1/custom/icds_reports/reports/awc_reports.py#L1131-L1132

Reply:
That example will perform two separate queries because the count is performed before the iteration. Ref [the source code for `count()`](https://github.com/django/django/blob/c00b4dbda2d9217cac205c4ffb3ccc1fa59a28ae/django/db/models/query.py#L388-L391)

Reply:
I think the point is that we don't need to compare the count separately if we're already comparing the full query